### PR TITLE
Add extension for ProcessArgumentBuilder

### PR DESCRIPTION
**Pull request by gep13**
So, instead of doing this:

```
builder.Append("-c");
builder.AppendQuoted(settings.CacheLocation);
```

You would be able to use:

```
builder.AppendSwitchQuoted("--version", settings.Version);
```

### CODE REVIEW

**Comment by gep13** [comment]
The work in this PR comes from @RichiCoder1, but in refactoring this PR https://github.com/cake-build/cake/pull/346 I removed it so that that PR only contained the work to get GitVersion included in Cake.  This separate PR adds the functionality back in.

**Comment by RichiCoder1** [comment]
I am Richicoder and I endorse this PR.

**Comment by gep13** [comment]
>@RichiCoder1 said...
I am Richicoder and I endorse this PR.

LOL! :smile_cat:

**Comment by patriksvensson** [comment]
This is nice I think:

```csharp
builder.AppendSwitchQuoted("--version", settings.Version);
```
But when using another separator, the syntax is a little weird (IMO):

```csharp
builder.AppendSwitchQuoted("--version", settings.Version, "=");
```

I think it would be much cleaner and intuitive if we added the separator between the switch and the value, so the expression will "read" the same way as it's written.
The downside is that we would need to introduce another overload.

```csharp
builder.AppendSwitchQuoted("--version", settings.Version);
builder.AppendSwitchQuoted("--version", "=", settings.Version);
```

What do you think? Worth it?

File: src/Cake.Core/IO/Arguments/SwitchArgument.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        /// <summary>
+        /// Render the arguments as a <see cref="System.String" />.
+        /// Sensitive information will be included.
+        /// </summary>
+        /// <returns>
+        /// A string representation of the argument.
+        /// </returns>
+        public string Render()
+        {
+            return string.Format(CultureInfo.InvariantCulture, "{0}{1}{2}", _switch, _seperator, _argument.Render());
```
Comment (patriksvensson) [inline src/Cake.Core/IO/Arguments/SwitchArgument.cs]: Is `Format` needed? Wouldn't it be better with `Concat` here?

File: src/Cake.Core/IO/Arguments/SwitchArgument.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        /// <summary>
+        /// Renders the argument as a <see cref="System.String" />.
+        ///  The secret text will be redacted.
+        /// </summary>
+        /// <returns>A safe string representation of the argument.</returns>
+        public string RenderSafe()
+        {
+            return string.Format(CultureInfo.InvariantCulture, "{0}{1}{2}", _switch, _seperator, _argument.RenderSafe());
```
Comment (patriksvensson) [inline src/Cake.Core/IO/Arguments/SwitchArgument.cs]: Is `Format` needed? Wouldn't it be better with `Concat` here?

File: src/Cake.Core/IO/Arguments/SwitchArgument.cs (hunk 1 above)
Comment (gep13) [inline src/Cake.Core/IO/Arguments/SwitchArgument.cs]: This doesn't bother me either way :+1: @RichiCoder1 if you have no objections, I will make the suggested change.

File: src/Cake.Core/IO/Arguments/SwitchArgument.cs (hunk 2 above)
Comment (gep13) [inline src/Cake.Core/IO/Arguments/SwitchArgument.cs]: This doesn't bother me either way :+1: @RichiCoder1 if you have no objections, I will make the suggested change.

**Comment by gep13** [comment]
@patriksvensson I agree, it is much easier to read with the additional overload.

So to confirm, remove the optional parameter from the first method call, and provide a second overload with a non optional parameter in the middle of the two others.

Yes?

**Comment by gep13** [comment]
@patriksvensson I have fixed up this PR with the suggestions that you made.

File: src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs [hunk 3]
```diff
@@ -177,5 +177,236 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
...
+
+        /// <summary>
+        /// Appends the specified switch to the argument builder.
+        /// </summary>
+        /// <param name="builder">The builder.</param>
+        /// <param name="switch">The switch preceding the text.</param>
+        /// <param name="seperator">The seperator between the switch and argument.</param>
+        /// <param name="text">The text to be appended.</param>
+        /// <returns>The same <see cref="ProcessArgumentBuilder"/> instance so that multiple calls can be chained.</returns>
+        public static ProcessArgumentBuilder AppendSwitch(this ProcessArgumentBuilder builder, string @switch, string seperator, string text)
```
Comment (patriksvensson) [inline src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs]: I suggest that this should call the overload that takes three arguments.

File: src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs [hunk 4]
```diff
@@ -177,5 +177,236 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
...
+        }
+
+        /// <summary>
+        /// Quotes and appends the specified text to the argument builder.
+        /// </summary>
+        /// <param name="builder">The builder.</param>
+        /// <param name="switch">The switch preceding the text.</param>
+        /// <param name="text">The text to be quoted and appended.</param>
+        /// <returns>The same <see cref="ProcessArgumentBuilder"/> instance so that multiple calls can be chained.</returns>
+        public static ProcessArgumentBuilder AppendSwitchQuoted(this ProcessArgumentBuilder builder, string @switch, string text)
```
Comment (patriksvensson) [inline src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs]: I suggest that this should call the overload that takes three arguments.

File: src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs [hunk 5]
```diff
@@ -177,5 +177,236 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
...
+        }
+
+        /// <summary>
+        /// Quotes and appends the specified argument to the argument builder.
+        /// </summary>
+        /// <param name="builder">The builder.</param>
+        /// <param name="switch">The switch preceding the text.</param>
+        /// <param name="argument">The argument to be quoted and appended.</param>
+        /// <returns>The same <see cref="ProcessArgumentBuilder"/> instance so that multiple calls can be chained.</returns>
+        public static ProcessArgumentBuilder AppendSwitchQuoted(this ProcessArgumentBuilder builder, string @switch, IProcessArgument argument)
```
Comment (patriksvensson) [inline src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs]: I suggest that this should call the overload that takes three arguments.

File: src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs [hunk 6]
```diff
@@ -177,5 +177,236 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
...
+        }
+
+        /// <summary>
+        /// Appends the specified secret text to the argument builder.
+        /// </summary>
+        /// <param name="builder">The builder.</param>
+        /// <param name="switch">The switch preceding the text.</param>
+        /// <param name="text">The secret text to be appended.</param>
+        /// <returns>The same <see cref="ProcessArgumentBuilder"/> instance so that multiple calls can be chained.</returns>
+        public static ProcessArgumentBuilder AppendSwitchSecret(this ProcessArgumentBuilder builder, string @switch, string text)
```
Comment (patriksvensson) [inline src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs]: I suggest that this should call the overload that takes three arguments.

File: src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs [hunk 7]
```diff
@@ -177,5 +177,236 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
...
+        }
+
+        /// <summary>
+        /// Appends the specified secret text to the argument builder.
+        /// </summary>
+        /// <param name="builder">The builder.</param>
+        /// <param name="switch">The switch preceding the text.</param>
+        /// <param name="argument">The secret argument to be appended.</param>
+        /// <returns>The same <see cref="ProcessArgumentBuilder"/> instance so that multiple calls can be chained.</returns>
+        public static ProcessArgumentBuilder AppendSwitchSecret(this ProcessArgumentBuilder builder, string @switch, IProcessArgument argument)
```
Comment (patriksvensson) [inline src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs]: I suggest that this should call the overload that takes three arguments.

File: src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs [hunk 8]
```diff
@@ -177,5 +177,236 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
...
+        }
+
+        /// <summary>
+        /// Quotes and appends the specified secret text to the argument builder.
+        /// </summary>
+        /// <param name="builder">The builder.</param>
+        /// <param name="switch">The switch preceding the text.</param>
+        /// <param name="text">The secret text to be quoted and appended.</param>
+        /// <returns>The same <see cref="ProcessArgumentBuilder"/> instance so that multiple calls can be chained.</returns>
+        public static ProcessArgumentBuilder AppendSwitchQuotedSecret(this ProcessArgumentBuilder builder, string @switch, string text)
```
Comment (patriksvensson) [inline src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs]: I suggest that this should call the overload that takes three arguments.

**Comment by RichiCoder1** [comment]
I think @patriksvensson suggests that these should call the overload that takes three arguments.

**Comment by patriksvensson** [comment]
@RichiCoder1 Haha :smile:

File: src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs (hunk 3 above)
Comment (gep13) [inline src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs]: Ha ha, it's all good :smile_cat:

**Comment by gep13** [comment]
@patriksvensson I believe that I have corrected up this PR with your suggestions.  Please let me know if I have done sorry incorrect.

**Comment by gep13** [comment]
@devlead @patriksvensson rebased and ready

**Comment by patriksvensson** [comment]
@gep13 Could you rebase this into a single commit?

**Comment by gep13** [comment]
@patriksvensson done

**Pull request merged**